pkg/sql/schemachanger/sctest/comparator.go

```diff
@@ -165,6 +182,142 @@ func modifyBlacklistedStmt(
 	return parsedLine.StringWithFlags(tree.FmtSimple | tree.FmtTagDollarQuotes)
 }
 
+// modifyAlterPKWithSamePKColsButDifferentSharding modifies any ALTER PK stmt in
+// `line` if the current/old primary index is hash-sharded and the new primary
+// index is also hash-sharded on the same key columns (but with a different
+// "bucket_count") by appending a `DROP COLUMN IF EXISTS
+// crdb_intenral_old_cols_shard` to it, so that legacy schema changer will
+// converge to declarative schema changer (in which ALTER PK will already drop
+// the old shard column).
+// The returned boolean indicates if such a modification happened.
+func modifyAlterPKWithSamePKColsButDifferentSharding(
+	ctx context.Context, t *testing.T, parsedStmts statements.Statements, conn *gosql.Conn,
+) (statements.Statements, bool) {
+	// getPKShardingInfo retrieve all sharding related information from table
+	// `name`'s PK.
+	getPKShardingInfo := func(name *tree.UnresolvedObjectName) (isSharded bool, shardColName string, shardBuckets int, columnNames []string) {
+		var shardingExists bool
+		row := conn.QueryRowContext(ctx, fmt.Sprintf(
+			`
+SELECT (crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor) -> 'table' -> 'primaryIndex' -> 'sharded') ? 'isSharded' AS sharded
```

Comment:
is there a way we can query this using pg_catalog or crdb_internal tables rather than needing to look at descriptors directly? (same question for the below queries)

Reply:
I don't know about pg_catalog (I haven't used any of the virtual tables in that schema). I know there are a few crdb_intenral builtin functions that returns schema information of descriptors but they are like `SHOW CREATE`. I find it easier to query the `system.descriptor` table, convert it to JSONB format, and access its fields.